File: app/src/main/java/com/depromeet/activity/GameActivity.java
package com.depromeet.activity;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.depromeet.R;
import com.depromeet.data.Poem;
import com.depromeet.data.WordResponse;
import com.depromeet.network.RetrofitBuilder;
import com.depromeet.network.ServiceApi;
import com.depromeet.util.FailDialog;
import com.depromeet.util.LoginManager;
import com.depromeet.util.PassDialog;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class GameActivity extends AppCompatActivity {
    private int mStartTime = 3;
    private int mQuizTime = 60;
    private String word = "";
    private LoginManager manager;
    private ServiceApi service;
    private TextView mStartTimerText;
    private TextView mQuizTimerText;
    private TextView mQuizFirstWordText;
    private TextView mQuizSecondWordText;
    private TextView mQuizThirdWordText;
    private EditText mQuizFirstAnswerEdit;
    private EditText mQuizSecondAnswerEdit;
    private EditText mQuizThirdAnswerEdit;
    private FloatingActionButton mSubmitBtn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game);
        manager = new LoginManager(this);
        service = RetrofitBuilder.INSTANCE.getInstance();

        mStartTimerText = findViewById(R.id.tv_game_beginCounter);
        mQuizTimerText = findViewById(R.id.tv_game_timer);
        mQuizFirstWordText = findViewById(R.id.tv_game_firstword);
        mQuizSecondWordText = findViewById(R.id.tv_game_secondword);
        mQuizThirdWordText = findViewById(R.id.tv_game_thirdword);
        mQuizFirstAnswerEdit = findViewById(R.id.et_game_first);
        mQuizSecondAnswerEdit = findViewById(R.id.et_game_second);
        mQuizThirdAnswerEdit = findViewById(R.id.et_game_third);
        mSubmitBtn = findViewById(R.id.fb_game_submit);
        requestRandomWord();

        // 5초 카운터
        new CountDownTimer(5000, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {  //1초마다 변하는 이벤트
                mQuizFirstAnswerEdit.setEnabled(false);
                mQuizSecondAnswerEdit.setEnabled(false);
                mQuizThirdAnswerEdit.setEnabled(false);

                mStartTimerText.setText(Integer.toString(--mStartTime));
            }

            @Override
            public void onFinish() {
                // 단어 셋팅
                mStartTimerText.setText(word);

                mQuizFirstWordText.setText(Character.toString(mStartTimerText.getText().charAt(0)));
                mQuizSecondWordText.setText(Character.toString(mStartTimerText.getText().charAt(1)));
                mQuizThirdWordText.setText(Character.toString(mStartTimerText.getText().charAt(2)));

                mQuizFirstAnswerEdit.setEnabled(true);
                mQuizSecondAnswerEdit.setEnabled(true);
                mQuizThirdAnswerEdit.setEnabled(true);

                // 5초 지난 후, 단어 등장과 함께 1분 카운트 시작
                new CountDownTimer(3000, 1000) {
                    @Override
                    public void onTick(long millisUntilFinished) {
                        mQuizTimerText.setText(Integer.toString(--mQuizTime));
                        if (!mQuizFirstAnswerEdit.getText().toString().equals("")) {
                            mQuizFirstWordText.setBackground(ContextCompat.getDrawable(GameActivity.this, R.drawable.ic_word));
                        }
                        if (!mQuizSecondAnswerEdit.getText().toString().equals("")) {
                            mQuizSecondWordText.setBackground(ContextCompat.getDrawable(GameActivity.this, R.drawable.ic_word));
                        }
                        if (!mQuizThirdAnswerEdit.getText().toString().equals("")) {
                            mQuizThirdWordText.setBackground(ContextCompat.getDrawable(GameActivity.this, R.drawable.ic_word));
                        }
                    }

                    @Override
                    public void onFinish() { //타이머 종료시 이벤트
                        // "노력해요" 팝업창
                        final FailDialog dialog = new FailDialog(GameActivity.this);
                        dialog.show();
                        // 2초후 메인화면 으로!
                        Handler handler = new Handler();
                        handler.postDelayed(new Runnable() {
                            public void run() {
                                dialog.dismiss();
                                finish();
                            }
                        }, 2000);
                    }
                }.start();
            }
        }.start();

        // 제출하기 버튼 누르면
        mSubmitBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if ((mQuizFirstAnswerEdit.getText().toString().length() != 0) &&
                        (mQuizSecondAnswerEdit.getText().toString().length() != 0) &&
                        (mQuizThirdAnswerEdit.getText().toString().length() != 0)) {
                    // 참!잘했어요 팝업창
                    showGoodDialog();
                }
            }
        });
    }

    private void showGoodDialog() {
        final PassDialog dialog = new PassDialog(GameActivity.this);
        dialog.show();

        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            public void run() {
                dialog.dismiss();
                finish();
            }
        }, 2000);
    }

    public void requestRandomWord() {
        service.getRandomWord().enqueue(new Callback<WordResponse>() {
            @Override
            public void onResponse(Call<WordResponse> call, Response<WordResponse> response) {
                word = response.body().getWord().getHangshi();
            }

            @Override
            public void onFailure(Call<WordResponse> call, Throwable t) {
                Toast.makeText(GameActivity.this, R.string.all_err, Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void requestSavePoem() {
        Poem poem = new Poem(
                manager.getUserId(),
                manager.getUserName(),
                mQuizFirstAnswerEdit.getText().toString(),
                mQuizSecondAnswerEdit.getText().toString(),
                mQuizThirdAnswerEdit.getText().toString(),
                0,
                false
        );

        service.savePoem(poem).enqueue(new Callback<com.depromeet.data.BasicResponse>() {
            @Override
            public void onResponse(Call<com.depromeet.data.BasicResponse> call, Response<com.depromeet.data.BasicResponse> response) {

            }

            @Override
            public void onFailure(Call<com.depromeet.data.BasicResponse> call, Throwable t) {

            }
        });
    }
}

File: app/src/main/java/com/depromeet/activity/GameActivity.kt
package com.depromeet.activity

import android.content.Context
import android.os.Bundle
import android.os.CountDownTimer
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.view.inputmethod.InputMethodManager
import com.depromeet.R
import com.depromeet.customView.GameResultDialog
import com.depromeet.data.BasicResponse
import com.depromeet.data.PoemRequest
import com.depromeet.data.WordResponse
import com.depromeet.network.RetrofitBuilder
import com.depromeet.network.ServiceApi
import com.depromeet.util.LoginManager
import kotlinx.android.synthetic.main.activity_game.*
import org.jetbrains.anko.alert
import org.jetbrains.anko.noButton
import org.jetbrains.anko.toast
import org.jetbrains.anko.yesButton
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class GameActivity : AppCompatActivity() {
    private val interval = 1000L
    private var waiting = 3L
    private var limitTime = 60L
    private var word = "삼행시"
    private lateinit var manager: LoginManager
    private lateinit var service: ServiceApi
    private lateinit var initTimer: CountDownTimer
    private lateinit var gameTimer: CountDownTimer

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)
        manager = LoginManager(this)
        service = RetrofitBuilder.getInstance()
        requestRandomWord()

        initView()
        initTimers()
        initTimer.start()
    }

    private fun initView() {
        tv_game_board.text = waiting.toString()
        tv_game_timer.text = limitTime.toString()
        fb_game_submit.setOnClickListener {
            if (TextUtils.isEmpty(et_game_word1.text)
                    || TextUtils.isEmpty(et_game_word2.text)
                    || TextUtils.isEmpty(et_game_word3.text)) {
                toast(R.string.game_not_input)
            } else {
                requestSavePoem()
            }
        }
    }

    private fun initTimers() {
        gameTimer = object : CountDownTimer(limitTime * interval, interval) {
            override fun onTick(millisUntilFinished: Long) {
                limitTime--
                tv_game_timer.text = limitTime.toString()
            }

            override fun onFinish() {
                showBadDialog()
            }
        }

        initTimer = object : CountDownTimer(waiting * interval, interval) {
            override fun onTick(millisUntilFinished: Long) {
                waiting--
                tv_game_board.text = waiting.toString()
            }

            override fun onFinish() {
                tv_game_board.text = word
                tv_game_word1.text = word.substring(0, 1)
                tv_game_word2.text = word.substring(1, 2)
                tv_game_word3.text = word.substring(2)
                et_game_word1.isEnabled = true
                et_game_word2.isEnabled = true
                et_game_word3.isEnabled = true

                // 첫번째 행에 키보드 포커스 할당
                et_game_word1.requestFocus()
                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.showSoftInput(et_game_word1, 0)

                gameTimer.start()
            }
        }
    }

    private fun showGoodDialog() {
        val dialog = GameResultDialog(this, true)
        dialog.show()
    }

    private fun showBadDialog() {
        fb_game_submit.isClickable = false
        et_game_word1.isEnabled = false
        et_game_word2.isEnabled = false
        et_game_word3.isEnabled = false
        val dialog = GameResultDialog(this, false)
        dialog.show()
    }

    private fun requestRandomWord() {
        service.getRandomWord().enqueue(object : Callback<WordResponse> {
            override fun onResponse(call: Call<WordResponse>, response: Response<WordResponse>) {
                word = response.body()!!.word.hangshi
            }

            override fun onFailure(call: Call<WordResponse>, t: Throwable) {
                toast(R.string.all_err)
            }
        })
    }

    private fun requestSavePoem() {
        val poem = PoemRequest(manager.userName, et_game_word1.text.toString(),
                et_game_word2.text.toString(), et_game_word3.text.toString())

        service.savePoem(poem).enqueue(object : Callback<BasicResponse> {
            override fun onResponse(call: Call<BasicResponse>, response: Response<BasicResponse>) {
                if (response.code() == 200) {
                    toast(R.string.game_finish)
                    showGoodDialog()
                } else {
                    toast(R.string.all_err)
                    finish()
                }
            }

            override fun onFailure(call: Call<BasicResponse>, t: Throwable) {
                toast(R.string.all_err)
            }
        })
    }

    override fun onBackPressed() {
        if (limitTime > 0) {
            alert(R.string.game_back) {
                yesButton { super.onBackPressed() }
                noButton {  }
            }.show()
        } else {
            super.onBackPressed()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            initTimer.cancel()
            gameTimer.cancel()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
